adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/providers/ChirpStackProvider.java

```
@@ -0,0 +1,147 @@
+
+    private static final String FIELD_CHIRPSTACK_PAYLOAD = "data";
+    private static final String FIELD_CHIRPSTACK_DEVICE = "devEUI";
+    private static final String FIELD_CHIRPSTACK_TX_INFO = "txInfo";
+    private static final String FIELD_CHIRPSTACK_SPREADING_FACTOR = "spreadingFactor";
+    private static final String FIELD_CHIRPSTACK_BANDWIDTH = "bandwidth";
+    private static final String FIELD_CHIRPSTACK_FUNCTION_PORT = "fPort";
+    private static final String FIELD_CHIRPSTACK_CODE_RATE = "codeRate";
+    private static final String FIELD_CHIRPSTACK_LORA_MODULATION_INFO = "loRaModulationInfo";
+    private static final String FIELD_CHIRPSTACK_FREQUENCY = "frequency";
+    private static final String FIELD_CHIRPSTACK_FRAME_COUNT = "fCnt";
+    private static final String FIELD_CHIRPSTACK_RX_INFO = "rxInfo";
+    private static final String FIELD_CHIRPSTACK_GATEWAY_ID = "gatewayID";
+    private static final String FIELD_CHIRPSTACK_RSSI = "rssi";
+    private static final String FIELD_CHIRPSTACK_LSNR = "loRaSNR";
+    private static final String FIELD_CHIRPSTACK_CHANNEL = "channel";
+    private static final String FIELD_CHIRPSTACK_LOCATION = "location";
+    private static final String FIELD_CHIRPSTACK_LATITUDE = "latitude";
+    private static final String FIELD_CHIRPSTACK_LONGITUDE = "longitude";
+
+    @Override
+    public String getProviderName() {
+        return "chirpStack";
+    }
+
+    @Override
+    public String pathPrefix() {
+        return "/chirpstack";
+    }
+
+    @Override
+    public String extractDeviceId(final JsonObject loraMessage) {
+        final Object deviceId = loraMessage.getValue(FIELD_CHIRPSTACK_DEVICE);
+        if (deviceId == null) {
+            throw new LoraProviderMalformedPayloadException("DeviceId could not be extracted from message.", null);
+        }
+        if (deviceId instanceof String) {
+            return LoraUtils.convertFromBase64ToHex((String) deviceId);
+        }
+
+        return null;
```

shouldn't we also throw the exception if the object is not null but is e.g. a number instead of a String? Isn't the *devEUI* field expected to be contained in each message and to hold a (Base64) String value?